Add ConfigFlow for seventeentrack integration

shaiu opened the pull request:
## Proposed change

- seventeentrack integration is now configured through config flow which will add the sensors automatically.
- Show archived/delivered packages are options (default False)

## Type of change
- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/31579

## Checklist
- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.


To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

shaiu committed:
Add config flow to 17Track

shaiu committed:
Import config from configuration.yaml

shaiu committed:
1. move import to async_setup_platform
2. add USERNAME (email) in title for uniqueness

shaiu committed:
Add options flow

shaiu committed:
Add tests

shaiu committed:
Add CONF_SHOW_ARCHIVED  and CONF_SHOW_DELIVERED to data from options

joostlek commented on homeassistant/components/seventeentrack/__init__.py:
```diff
@@ -1 +1,68 @@
+    """Set up the 17Track component."""
+    hass.data.setdefault(DOMAIN, {})
+
+    if hass.config_entries.async_entries(DOMAIN):
+        return True
```
We should create repair issues instead to point users to the fact that they need to remove the YAML. Please checkout integrations like `lutron` and `suez_water` for this.

shaiu replied:
Done

joostlek commented on homeassistant/components/seventeentrack/__init__.py:
```diff
@@ -1 +1,68 @@
+
+async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
+    """Set up the 17Track component."""
+    hass.data.setdefault(DOMAIN, {})
```
```suggestion
```

joostlek commented on homeassistant/components/seventeentrack/__init__.py:
```diff
@@ -1 +1,68 @@
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up 17Track from a config entry."""
+
+    hass.data.setdefault(DOMAIN, {})
+
+    session = async_get_clientsession(hass)
+    client = SeventeenTrackClient(session=session)
+
+    try:
+        await client.profile.login(entry.data[CONF_USERNAME], entry.data[CONF_PASSWORD])
+    except SeventeenTrackError as err:
+        raise ConfigEntryNotReady from err
+
+    hass.data[DOMAIN][entry.entry_id] = client
```
```suggestion
    session = async_get_clientsession(hass)
    client = SeventeenTrackClient(session=session)

    try:
        await client.profile.login(entry.data[CONF_USERNAME], entry.data[CONF_PASSWORD])
    except SeventeenTrackError as err:
        raise ConfigEntryNotReady from err

    hass.data.setdefault(DOMAIN, {})[entry.entry_id] = client
```

joostlek commented on homeassistant/components/seventeentrack/config_flow.py:
```diff
@@ -0,0 +1,105 @@
+        """Get options flow for this handler."""
+        return SchemaOptionsFlowHandler(config_entry, OPTIONS_FLOW)
+
+    async def _async_validate_input(self, user_input):
+        """Validate the user input allows us to connect."""
+
+        session = aiohttp_client.async_get_clientsession(self.hass)
+
+        client = SeventeenTrackClient(session=session)
+
+        try:
+            login_result = await client.profile.login(
+                user_input[CONF_USERNAME], user_input[CONF_PASSWORD]
+            )
+
+            if not login_result:
+                _LOGGER.error("Invalid username and password provided")
+                return {"base": "invalid_credentials"}
+        except SeventeenTrackError as err:
+            _LOGGER.error("There was an error while logging in: %s", err)
+            return {"base": "cannot_connect"}
+
+        return {}
```
I think this function would work better inlined in the user step

joostlek commented on homeassistant/components/seventeentrack/config_flow.py:
```diff
@@ -0,0 +1,105 @@
+                user_input[CONF_USERNAME], user_input[CONF_PASSWORD]
+            )
+
+            if not login_result:
+                _LOGGER.error("Invalid username and password provided")
+                return {"base": "invalid_credentials"}
```
Only have stuff in the try block that can raise

joostlek commented on homeassistant/components/seventeentrack/config_flow.py:
```diff
@@ -0,0 +1,105 @@
+        """Handle a flow initialized by the user."""
+
+        errors = {}
+        if user_input is not None:
```
```suggestion
        if user_input:
```

joostlek commented on homeassistant/components/seventeentrack/config_flow.py:
```diff
@@ -0,0 +1,105 @@
+
+        return {}
+
+    async def async_step_user(
```
This step is missing a check for uniqueness. Can we fetch some kind of value from 17track that could act as unique identifier? Otherwise we should use `async_abort_entries_match` to compare the new user input with existing entities.

Can users change their username in the 17track app?

joostlek commented on homeassistant/components/seventeentrack/manifest.json:
```diff
@@ -1,8 +1,10 @@
+  "codeowners": ["@shaiu"],
+  "config_flow": true,
   "documentation": "https://www.home-assistant.io/integrations/seventeentrack",
+  "integration_type": "hub",
```
```suggestion
  "integration_type": "service",
```
I'd argue that this is a service instead of a hub

joostlek commented on homeassistant/components/seventeentrack/__init__.py:
```diff
@@ -1 +1,68 @@
+
+PLATFORMS = [Platform.SENSOR]
+
+CONFIG_SCHEMA = vol.Schema(
+    {
+        DOMAIN: vol.Schema(
+            {
+                vol.Required(CONF_USERNAME): str,
+                vol.Required(CONF_PASSWORD): str,
+                vol.Optional(CONF_SHOW_ARCHIVED, default=False): bool,
+                vol.Optional(CONF_SHOW_DELIVERED, default=False): bool,
+            },
+        )
+    },
+    extra=vol.ALLOW_EXTRA,
+)
```
We should keep this import in the `sensor.py`

shaiu replied:
I'm not sure what you mean?
If I move this to `sensor.py` I get the following error:
`[ERROR] [CONFIG_SCHEMA] Integrations which implement 'async_setup' or 'setup' must define either 'CONFIG_SCHEMA', 'PLATFORM_SCHEMA' or 'PLATFORM_SCHEMA_BASE'`

joostlek replied:
Currently the integration is setup in sensor.py (with I think either (async)setup(platform))

shaiu replied:
removed setup from `sensor.py`

joostlek commented on homeassistant/components/seventeentrack/sensor.py:
```diff
@@ -95,14 +86,32 @@ async def async_setup_platform(
+
+    client = hass.data[DOMAIN][entry_id]
+
+    scan_interval = config_entry.options.get(CONF_SCAN_INTERVAL, DEFAULT_SCAN_INTERVAL)
```
Currently this is never set or imported afaict

shaiu replied:
see Line 314:
`self.async_update = Throttle(self._scan_interval)(self._async_update)`

joostlek replied:
yes, but I am talking about how this data got into `config_entry.options`

shaiu commented:
@joostlek thanks for the quick review. I'll work on it now.

shaiu committed:
Update homeassistant/components/seventeentrack/__init__.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

shaiu committed:
Update homeassistant/components/seventeentrack/__init__.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

shaiu committed:
Update homeassistant/components/seventeentrack/config_flow.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

shaiu committed:
Update homeassistant/components/seventeentrack/manifest.json

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

shaiu committed:
1. Added repair issues
2. _async_validate_input inlined
3. added unique id
4. take default scan interval

CFenner commented on homeassistant/components/seventeentrack/config_flow.py:
```diff
@@ -0,0 +1,105 @@
+
+            if not errors:
+                return self.async_create_entry(
+                    title=f"17Track {user_input[CONF_USERNAME]}", data=user_input
```
I would omit the integration name in the entry name, it adds no real information.

shaiu committed:
Update homeassistant/components/seventeentrack/config_flow.py

Co-authored-by: Christopher Fenner <9592452+CFenner@users.noreply.github.com>

shaiu committed:
Update homeassistant/components/seventeentrack/__init__.py

Co-authored-by: Christopher Fenner <9592452+CFenner@users.noreply.github.com>

shaiu committed:
Update homeassistant/components/seventeentrack/sensor.py

Co-authored-by: Christopher Fenner <9592452+CFenner@users.noreply.github.com>

CFenner commented on homeassistant/components/seventeentrack/__init__.py:
```diff
@@ -1 +1,111 @@
+
+CONFIG_SCHEMA = vol.Schema(
+    {
+        DOMAIN: vol.Schema(
```
Is it possible to combine `USER_SCHEMA` and `OPTIONS_SCHEMA` here to avoid code duplication?

shaiu replied:
Done

shaiu committed:
fix

shaiu committed:
1. move async_create_issue to async_setup_platform
2. fix tests
3. black + pylint

shaiu committed:
combine USER_SCHEMA and OPTIONS_SCHEMA

shaiu committed:
small fix

shaiu committed:
remove async_setup

shaiu committed:
fix tests and add 100% coverage

shaiu commented:
@joostlek @CFenner I finished fixing all comments + added more test with more coverage.
Thanks.

joostlek commented:
Will take a look later :)

joostlek commented on homeassistant/components/seventeentrack/__init__.py:
```diff
@@ -1 +1,41 @@
+
+PLATFORMS = [Platform.SENSOR]
+
+CONFIG_SCHEMA = vol.Schema(
+    {
+        DOMAIN: USER_SCHEMA,
+    },
+    extra=vol.ALLOW_EXTRA,
+)
```
Can be removed

shaiu replied:
removed

joostlek commented on homeassistant/components/seventeentrack/config_flow.py:
```diff
@@ -0,0 +1,124 @@
+                if not await client.profile.login(
+                    user_input[CONF_USERNAME], user_input[CONF_PASSWORD]
+                ):
+                    _LOGGER.error("Invalid username and password provided")
```
This error doesn't add anything as the error message already reflects this

shaiu replied:
removed

joostlek commented on homeassistant/components/seventeentrack/sensor.py:
```diff
@@ -77,32 +49,43 @@ async def async_setup_platform(
-
-    client = SeventeenTrackClient(session=session)
+    """Initialize 17Track import from config."""
+    async_create_issue(
+        hass,
+        HOMEASSISTANT_DOMAIN,
+        f"deprecated_yaml_{DOMAIN}",
+        is_fixable=False,
+        breaks_in_ha_version="2024.7.0",
+        severity=IssueSeverity.WARNING,
+        translation_key="deprecated_yaml",
+        translation_placeholders={
+            "domain": DOMAIN,
+            "integration_title": "17Track",
+        },
+    )
```
Issues should be guiding the user more to what they did wrong. Did the connection fail? Are their credentials incorrect? Please check integrations like `lutron`, `suez_water` or `streamlabswater` for examples

shaiu replied:
updated

joostlek commented on homeassistant/components/seventeentrack/config_flow.py:
```diff
@@ -0,0 +1,124 @@
+            errors=errors,
+        )
+
+    async def async_step_import(self, import_data) -> FlowResult:
```
```suggestion
    async def async_step_import(self, import_data: dict[str, Any]) -> FlowResult:
```

joostlek reviewed (changes requested):
A quick glance

shaiu committed:
1. remove CONFIG_SCHEMA
2. remove error log
3. add issue with more description when import issues happen
4. some linting

shaiu committed:
Update homeassistant/components/seventeentrack/config_flow.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

shaiu commented:
@joostlek Done

joostlek commented:
Please don't separately ping people, when you press ready for review the right people are notified again :)

joostlek commented on homeassistant/components/seventeentrack/sensor.py:
```diff
@@ -78,32 +60,57 @@ async def async_setup_platform(
+            HOMEASSISTANT_DOMAIN,
+            f"deprecated_yaml_{DOMAIN}",
+            is_fixable=False,
+            breaks_in_ha_version="2024.7.0",
```
```suggestion
            breaks_in_ha_version="2024.9.0",
```
Deprecation period is 6 months and this will be introduced in at least 2024.3

joostlek commented on homeassistant/components/seventeentrack/sensor.py:
```diff
@@ -78,32 +60,57 @@ async def async_setup_platform(
+            hass,
+            DOMAIN,
+            f"deprecated_yaml_import_issue_${result['reason']}",
+            breaks_in_ha_version="2024.7.0",
```
```suggestion
            breaks_in_ha_version="2024.9.0",
```

joostlek commented on homeassistant/components/seventeentrack/sensor.py:
```diff
@@ -126,12 +133,14 @@ def __init__(self, data, status, initial_state):
-    def native_value(self):
+    def native_value(
+        self,
+    ) -> StateType | str | int | float | None | date | datetime | Decimal:
```
```suggestion
    ) -> StateType:
```
I think this should be enough

joostlek commented on homeassistant/components/seventeentrack/sensor.py:
```diff
@@ -195,14 +204,16 @@ def available(self) -> bool:
 
     @property
-    def name(self):
+    def name(self) -> str | UndefinedType | None:
```
This always returns a string

joostlek commented on homeassistant/components/seventeentrack/sensor.py:
```diff
@@ -195,14 +204,16 @@ def available(self) -> bool:
-    def native_value(self):
+    def native_value(
+        self,
+    ) -> StateType | str | int | float | None | date | datetime | Decimal:
```
Only state type

shaiu replied:
changed to only StateType

joostlek commented on tests/components/seventeentrack/__init__.py:
```diff
@@ -1 +1,150 @@
+}
+
+
+async def _goto_future(hass, future=None):
+    """Move to future."""
+    if not future:
+        future = utcnow() + datetime.timedelta(minutes=10)
+    with patch("homeassistant.util.utcnow", return_value=future):
+        async_fire_time_changed(hass, future)
+        await hass.async_block_till_done()
+        async_fire_time_changed(hass, future)
+        await hass.async_block_till_done()
+
```
We use the freezer for these tasks. We can do `freezer.tick(..)` and it will skip time (should be combined with async fire time changed).

joostlek commented on tests/components/seventeentrack/__init__.py:
```diff
@@ -1 +1,150 @@
+        tz: str = "UTC",
+    ) -> list:
+        """Packages mock."""  # noqa: D401
+        if self.__class__.package_error:
+            raise SeventeenTrackError("Error")
+        return self.__class__.package_list[:]
+
+    async def summary(self, show_archived: bool = False) -> dict:
+        """Summary mock."""
+        if self.__class__.summary_error:
+            raise SeventeenTrackError("Error")
+        return self.__class__.summary_data
+
+
+class ClientMock:
+    """Mock the py17track client to inject the ProfileMock."""
+
+    def __init__(self, session) -> None:
+        """Mock the profile."""
+        self.profile = ProfileMock()
+
+
+async def init_integration(hass, config, options=None) -> MockConfigEntry:
+    """Set up the 17Track integration in Home Assistant."""
+
+    if options is None:
+        options = {}
+    entry = MockConfigEntry(
+        domain=DOMAIN,
+        title="17Track",
+        unique_id="0123456",
+        data=config,
+        options=options,
+    )
+
+    entry.add_to_hass(hass)
+    await hass.config_entries.async_setup(entry.entry_id)
+    await hass.async_block_till_done()
+
+    return entry
```
Preferably we use normal mocks for these things (as in, an AsyncMock or MagicMock).

Tests I really recommend checking are `epion` or `analytics_insights`

shaiu replied:
using AsyncMock now

joostlek commented on tests/components/seventeentrack/test_config_flow.py:
```diff
@@ -0,0 +1,166 @@
+    assert result["errors"] == {}
+
+    with patch(
+        "py17track.profile.Profile.login",
```
Please patch libraries at the places where they are used (homeassistant.components.seventeentrack...)

shaiu replied:
Done

joostlek commented on tests/components/seventeentrack/test_config_flow.py:
```diff
@@ -0,0 +1,166 @@
+        }
+
+
+async def test_invalid_cred_error(hass: HomeAssistant) -> None:
+    """Test that the user step fails."""
+    with patch(
+        "py17track.profile.Profile.login",
+        return_value=False,
+    ):
+        result = await hass.config_entries.flow.async_init(
+            DOMAIN,
+            context={"source": SOURCE_USER},
+            data=VALID_CONFIG,
+        )
+
+        assert result["errors"] == {"base": "invalid_credentials"}
+
+
+async def test_cannot_connect_error(hass: HomeAssistant) -> None:
+    """Test that the user step fails."""
+    with patch(
+        "py17track.profile.Profile.login",
+        side_effect=SeventeenTrackError(),
+    ):
+        result = await hass.config_entries.flow.async_init(
+            DOMAIN,
+            context={"source": SOURCE_USER},
+            data=VALID_CONFIG,
+        )
+
+        assert result["errors"] == {"base": "cannot_connect"}
```
We should test if the config flow is able to recover from an error, so please patch everything to work again and make it succeed. 

This can be parametrized, I think `lutron` and `lastfm` can show this well.

shaiu replied:
Done and Done

joostlek commented on tests/components/seventeentrack/test_config_flow.py:
```diff
@@ -0,0 +1,166 @@
+        assert result["data"][CONF_PASSWORD] == "edc3eee7330e4fdda04489e3fbc283d0"
+
+
+async def test_import_flow_cannot_connect_error(hass: HomeAssistant) -> None:
+    """Test the import configuration flow with cannot_connect error."""
+    with patch(
+        "py17track.profile.Profile.login",
+        side_effect=SeventeenTrackError(),
+    ):
+        result = await hass.config_entries.flow.async_init(
+            DOMAIN,
+            context={"source": SOURCE_IMPORT},
+            data=VALID_CONFIG_OLD,
+        )
+
+        assert result["reason"] == "cannot_connect"
+
+
+async def test_import_flow_invalid_cred_error(hass: HomeAssistant) -> None:
+    """Test the import configuration flow with cannot_connect error."""
+    with patch(
+        "py17track.profile.Profile.login",
+        return_value=False,
+    ):
+        result = await hass.config_entries.flow.async_init(
+            DOMAIN,
+            context={"source": SOURCE_IMPORT},
+            data=VALID_CONFIG_OLD,
+        )
+
+        assert result["reason"] == "invalid_credentials"
```
Can also use parametrize

shaiu replied:
Done

joostlek commented on tests/components/seventeentrack/test_config_flow.py:
```diff
@@ -0,0 +1,166 @@
+"""Define tests for the 17Track config flow."""
```
I'm missing tests for the unique id. So please add a test where you first add a mock config entry with the same unique and set up the integration, it should abort

shaiu replied:
Added a test for aborting when configuring same unique id

joostlek reviewed (changes requested):
Maybe it makes sense to improve the tests in a preliminary PR

shaiu committed:
Update homeassistant/components/seventeentrack/sensor.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

shaiu committed:
Update homeassistant/components/seventeentrack/sensor.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

shaiu committed:
Update homeassistant/components/seventeentrack/sensor.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

joostlek commented:
Oh and please rebase on dev, there are some new typing changes in the config flow that this PR should adhere to

shaiu committed:
use freezer
use AsyncMock
fix tests

shaiu committed:
add test_flow_fails
parametrize tests where needed
test_import_flow_already_configured - where a unique id already configured (abort flow)

shaiu committed:
lint

shaiu committed:
fix rebase issues

shaiu committed:
some more fix

shaiu commented:
> Maybe it makes sense to improve the tests in a preliminary PR

So you want me to open a separate PR for that? Although I already fixed all issues in this PR already

joostlek commented:
It keeps the PRs more pure, this way we can focus on the tests in that PR and focus on the config flow in this PR

shaiu committed:
17Track revert tests and put them in a different PR

joostlek commented:
Hold up, not saying you should revert the old changes, you can just create a new branch from dev and include the tests. Since the old way of setting up is a bit different you would still need to change stuff eventually, but I think fixing the tests so they use an easier way to mock and use a better function to set up the integration would be nice to add in a PR we will merge before this one, and then you can incorporate these changes in this PR

shaiu commented:
> Hold up, not saying you should revert the old changes, you can just create a new branch from dev and include the tests. Since the old way of setting up is a bit different you would still need to change stuff eventually, but I think fixing the tests so they use an easier way to mock and use a better function to set up the integration would be nice to add in a PR we will merge before this one, and then you can incorporate these changes in this PR

Now I got what you mean.. 😄

shaiu committed:
adapt tests to MockConfigEntry

joostlek commented on homeassistant/components/seventeentrack/sensor.py:
```diff
@@ -72,39 +59,80 @@
     }
 )
 
+CONFIG_SCHEMA = vol.Schema(
+    {
+        DOMAIN: vol.Schema(
+            {
+                vol.Required(CONF_USERNAME): str,
+                vol.Required(CONF_PASSWORD): str,
+                vol.Optional(CONF_SHOW_ARCHIVED, default=False): bool,
+                vol.Optional(CONF_SHOW_DELIVERED, default=False): bool,
+            },
+        )
+    },
+    extra=vol.ALLOW_EXTRA,
+)
```
This should not be meeded

shaiu replied:
removed

joostlek commented on homeassistant/components/seventeentrack/sensor.py:
```diff
@@ -72,39 +59,80 @@
+            HOMEASSISTANT_DOMAIN,
+            f"deprecated_yaml_{DOMAIN}",
+            is_fixable=False,
+            breaks_in_ha_version="2024.9.0",
```
```suggestion
            breaks_in_ha_version="2024.10.0",
```

joostlek commented on homeassistant/components/seventeentrack/sensor.py:
```diff
@@ -72,39 +59,80 @@
+            hass,
+            DOMAIN,
+            f"deprecated_yaml_import_issue_${result['reason']}",
+            breaks_in_ha_version="2024.9.0",
```
```suggestion
            breaks_in_ha_version="2024.10.0",
```

joostlek reviewed (changes requested):
Almost there :)

shaiu committed:
Update tests/components/seventeentrack/test_sensor.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

shaiu committed:
Update homeassistant/components/seventeentrack/sensor.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

shaiu committed:
Update homeassistant/components/seventeentrack/sensor.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

shaiu commented on tests/components/seventeentrack/conftest.py:
```diff
@@ -62,14 +35,55 @@
 def mock_seventeentrack():
     """Build a fixture for the 17Track API."""
-    mock_seventeentrack_api = AsyncMock()
+    mock_seventeentrack_profile = AsyncMock(account_id=ACCOUNT_ID)
+    mock_seventeentrack_api = AsyncMock(profile=mock_seventeentrack_profile)
```
This is in order to have the mock 17track API return ACCOUNT_ID in order to test the unique id config flow. That's how I set the unique id for the config flow

joostlek replied:
Oooh right. but wouldn't it make more sense to have this in the `with patch....` block? Just like the other patches?

shaiu replied:
right.
Done

shaiu committed:
Update tests/components/seventeentrack/__init__.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

shaiu committed:
1. create fixture for config and another with options
2. set options with default values
3. remove CONFIG_SCHEMA

joostlek commented on tests/components/seventeentrack/conftest.py:
```diff
@@ -62,14 +41,80 @@
+        yield mock_setup_entry
+
+
+@pytest.fixture(autouse=True)
```
```suggestion
@pytest.fixture
```

joostlek commented on tests/components/seventeentrack/conftest.py:
```diff
@@ -62,14 +41,80 @@
+    )
+
+
+@pytest.fixture(autouse=True)
```
```suggestion
@pytest.fixture
```

joostlek commented on tests/components/seventeentrack/conftest.py:
```diff
@@ -62,14 +41,80 @@
+
+
+@pytest.fixture(autouse=True)
+def mock_config_entry_no_options() -> MockConfigEntry:
```
Perhaps we need to pick a better name

shaiu replied:
renamed

joostlek commented on homeassistant/components/seventeentrack/config_flow.py:
```diff
@@ -0,0 +1,133 @@
+
+        await self.async_set_unique_id(account_id)
+        self._abort_if_unique_id_configured()
+        return self.async_create_entry(
+            title=import_data[CONF_USERNAME],
+            data=import_data,
+            options={
+                CONF_SHOW_ARCHIVED: DEFAULT_SHOW_ARCHIVED,
+                CONF_SHOW_DELIVERED: DEFAULT_SHOW_DELIVERED,
+            },
+        )
```
For the options in the import: dont forget people also have these fields set, so we should use that field from the YAML, otherwise set to default

shaiu replied:
Done

shaiu committed:
Update tests/components/seventeentrack/conftest.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

shaiu committed:
Update tests/components/seventeentrack/conftest.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

shaiu committed:
1. get options from import data and default if not present
2. rename mock_config_entry_no_options -> mock_config_entry_with_default_options

shaiu commented on homeassistant/components/seventeentrack/config_flow.py:
```diff
@@ -0,0 +1,123 @@
+                self._abort_if_unique_id_configured()
+
+                return self.async_create_entry(
+                    title=user_input[CONF_USERNAME], data=user_input
```
Done

shaiu committed:
move ACCOUNT_ID to mock_seventeentrack_api.return_value.profile.account_id

joostlek commented on homeassistant/components/seventeentrack/config_flow.py:
```diff
@@ -0,0 +1,137 @@
+                if not await client.profile.login(
+                    user_input[CONF_USERNAME], user_input[CONF_PASSWORD]
+                ):
+                    errors = {"base": "invalid_credentials"}
```
```suggestion
                    errors["base"] = "invalid_auth"
```

joostlek commented on homeassistant/components/seventeentrack/config_flow.py:
```diff
@@ -0,0 +1,137 @@
+                    errors = {"base": "invalid_credentials"}
+            except SeventeenTrackError as err:
+                _LOGGER.error("There was an error while logging in: %s", err)
+                errors = {"base": "cannot_connect"}
```
```suggestion
                errors["base"] = "cannot_connect"
```

joostlek commented on homeassistant/components/seventeentrack/sensor.py:
```diff
@@ -127,12 +141,14 @@ def __init__(self, data, status, initial_state):
 
     @property
-    def native_value(self):
+    def native_value(
+        self,
+    ) -> StateType:
```
```suggestion
    def native_value(self) -> StateType:
```

joostlek commented on homeassistant/components/seventeentrack/sensor.py:
```diff
@@ -196,14 +212,16 @@ def available(self) -> bool:
 
     @property
-    def native_value(self):
+    def native_value(
+        self,
+    ) -> StateType:
```
```suggestion
    def native_value(self) -> StateType:
```

joostlek commented on homeassistant/components/seventeentrack/strings.json:
```diff
@@ -0,0 +1,41 @@
+    },
+    "error": {
+      "already_configured": "[%key:common::config_flow::abort::already_configured_device%]",
+      "invalid_credentials": "[%key:common::config_flow::error::invalid_auth%]",
```
```suggestion
      "invalid_auth": "[%key:common::config_flow::error::invalid_auth%]",
```

joostlek commented on homeassistant/components/seventeentrack/strings.json:
```diff
@@ -0,0 +1,41 @@
+      "already_configured": "[%key:common::config_flow::abort::already_configured_device%]"
+    },
+    "error": {
+      "already_configured": "[%key:common::config_flow::abort::already_configured_device%]",
```
```suggestion
```

joostlek commented on homeassistant/components/seventeentrack/strings.json:
```diff
@@ -0,0 +1,41 @@
+      }
+    },
+    "abort": {
+      "already_configured": "[%key:common::config_flow::abort::already_configured_device%]"
```
```suggestion
      "already_configured": "[%key:common::config_flow::abort::already_configured_service%]"
```

joostlek commented on homeassistant/components/seventeentrack/strings.json:
```diff
@@ -0,0 +1,41 @@
+    },
+    "deprecated_yaml_import_issue_invalid_credentials": {
+      "title": "The 17Track YAML configuration import request failed due to invalid credentials",
+      "description": "Configuring 17Track using YAML is being removed but there was an invalid credentials error while importing your existing configuration.\nSetup will not proceed.\n\nVerify that your 17Track credentials are correct and restart Home Assistant to attempt the import again.\n\nAlternatively, you may remove the 17Track configuration from your YAML configuration entirely, restart Ho ...
```
```suggestion
      "description": "Configuring 17Track using YAML is being removed but there were invalid credentials provided while importing your existing configuration.\nSetup will not proceed.\n\nVerify that your 17Track credentials are correct and restart Home Assistant to attempt the import again.\n\nAlternatively, you may remove the 17Track configuration from your YAML configuration entirely, restart Home Assistant, and add the 17Track integration manually."
```

joostlek commented on homeassistant/components/seventeentrack/strings.json:
```diff
@@ -0,0 +1,41 @@
+  "issues": {
+    "deprecated_yaml_import_issue_cannot_connect": {
+      "title": "The 17Track YAML configuration import cannot connect to server",
+      "description": "Configuring 17Track using YAML is being removed but there was an connection error importing your YAML configuration.\n\nThings you can try:\nMake sure your home assistant can reach the web.\n\nThen restart Home Assistant to try importing this integration again.\n\nAlternatively, you may remove the 17Track configuration from your YAML configuration entirely, restart Home Assi ...
```
```suggestion
      "description": "Configuring 17Track using YAML is being removed but there was a connection error importing your YAML configuration.\n\nThings you can try:\nMake sure your home assistant can reach the web.\n\nThen restart Home Assistant to try importing this integration again.\n\nAlternatively, you may remove the 17Track configuration from your YAML configuration entirely, restart Home Assistant, and add the 17Track integration manually."
```

joostlek commented on tests/components/seventeentrack/test_config_flow.py:
```diff
@@ -0,0 +1,205 @@
+        context={"source": SOURCE_IMPORT},
+        data=VALID_CONFIG_OLD,
+    )
+
+    assert result["reason"] == error
```
```suggestion

    assert result["type"] == data_entry_flow.FlowResultType.ABORT
    assert result["reason"] == error
```

joostlek commented on tests/components/seventeentrack/test_config_flow.py:
```diff
@@ -0,0 +1,205 @@
+
+    await hass.config_entries.async_setup(entry.entry_id)
+    await hass.async_block_till_done()
+    result = await hass.config_entries.options.async_init(
+        entry.entry_id,
+        data=None,
+    )
```
```suggestion
    result = await hass.config_entries.options.async_init(entry.entry_id, data={})
```

joostlek commented on tests/components/seventeentrack/test_config_flow.py:
```diff
@@ -0,0 +1,205 @@
+
+async def test_option_flow(hass: HomeAssistant, mock_seventeentrack: AsyncMock) -> None:
+    """Test option flow."""
+    entry = MockConfigEntry(domain=DOMAIN, data=VALID_CONFIG)
```
```suggestion
    entry = MockConfigEntry(
        domain=DOMAIN,
        data=VALID_CONFIG,
        options={
            CONF_SHOW_ARCHIVED: False,
            CONF_SHOW_DELIVERED: False,
        },
    )
```

joostlek commented on tests/components/seventeentrack/test_config_flow.py:
```diff
@@ -0,0 +1,205 @@
+    entry = MockConfigEntry(domain=DOMAIN, data=VALID_CONFIG)
+    entry.add_to_hass(hass)
+
+    assert not entry.options
+
```
```suggestion
```

joostlek committed:
Apply suggestions from code review

joostlek committed:
Merge branch 'dev' into shaiu-17-track-config-flow

joostlek commented on tests/components/seventeentrack/test_config_flow.py:
```diff
@@ -0,0 +1,208 @@
+        (
+            False,
+            None,
+            "invalid_credentials",
```
```suggestion
            "invalid_auth",
```

joostlek committed:
Update tests/components/seventeentrack/test_config_flow.py

shaiu commented:
@joostlek Thank you very much for your patience of guiding and reviewing this for so long.
I really appreciate it and learned a lot from the process.

joostlek commented:
Feel free to hit me up on Discord, I see some more points for improvement within this integration and wanted to check them with you :)

MartinHjelmare commented on homeassistant/components/seventeentrack/config_flow.py:
```diff
@@ -0,0 +1,137 @@
+            },
+        )
+
+    async def _get_client(self):
```
This doesn't need to be a coroutine function since we don't await inside. Decorate it with `@callback` instead from core.py.

shaiu replied:
Done

MartinHjelmare commented on homeassistant/components/seventeentrack/sensor.py:
```diff
@@ -72,39 +59,66 @@
+        async_create_issue(
+            hass,
+            DOMAIN,
+            f"deprecated_yaml_import_issue_${result['reason']}",
```
Why is there a dollar sign in the string?

shaiu replied:
removed

MartinHjelmare commented on homeassistant/components/seventeentrack/strings.json:
```diff
@@ -0,0 +1,40 @@
+    },
+    "abort": {
+      "already_configured": "[%key:common::config_flow::abort::already_configured_service%]"
+    },
```
We're missing abort reasons for cannot_connect and invalid_credentials.

shaiu replied:
added

MartinHjelmare reviewed (commented):
Please address the comments in a new PR. Thanks!

shaiu commented:
@MartinHjelmare https://github.com/home-assistant/core/pull/114283
Thanks
